src/cursor/common.rs

```diff
@@ -88,6 +103,18 @@ where
                     Poll::Ready(get_more_result) => {
                         let exhausted = get_more_result.exhausted();
                         let (result, session) = get_more_result.into_parts();
+                        if exhausted {
+                            // If the cursor is exhausted, the driver must return the pinned
+                            // connection to the pool.
+                            self.pinned_connection = PinnedConnection::new(None);
+                        }
+                        if let Err(e) = &result {
+                            if e.is_network_error() {
+                                // Flag the connection as invalid, preventing a killCursors command,
+                                // but leave the connection pinned.
+                                self.pinned_connection.invalidate();
```

Comment:
Once the connection is invalidated, we will probably want to return `Poll::Ready(None)` in all subsequent polls after returning the error.

We don't currently handle this properly in the regular code path either (see the `?` below without any additional handling). We can fix that if necessary in a separate piece of work though.

Reply:
Updated the poll to return `Poll::Ready(None)` for an invalidated connection; would prefer to leave handling general errors to a different PR.